Update Lit Actions Advanced Examples (#332)
LIT-Protocol/docs: docs/sdk/serverless-signing/combining-signatures.md

import FeedbackComponent from "@site/src/pages/feedback.md";

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Signing within a Lit Action

## Overview

When [signing a transaction](../serverless-signing/quick-start#sign-a-transaction.md) with Lit, signature shares are typically combined client-side. However, the `signAndCombineEcdsa` function allows you to combine signature shares directly within a Lit Action, which is useful for when you want to make use of the signed data within your Lit Action e.g. submitting a signed transaction. The signature shares will remain within the confines of each Lit node's [Trusted Execution Environment (TEE)](../../resources/how-it-works#1-lit-nodes.md) without ever being exposed to the outside world. 

> When you call the `signAndCombineEcdsa` function, signature shares are collected from each Lit node before being combined on a *single* node. The following code example will show how you can use this functionality for signing a blockchain transaction using ethers.js.

Reviewer: > The following code example will show how you can use this functionality for signing a blockchain transaction using ethers.js.

consider "The following example demonstrates how you can use this functionality to sign a blockchain transaction using ethers.js"

## Prerequisites

- Knowlege of [SessionSigs](../authentication/session-sigs/intro)
- Basic understanding of [Lit Actions](../serverless-signing/quick-start)

## Complete Code Example

The complete code example is available in the [Lit Developer Guides Code Repository](https://github.com/LIT-Protocol/developer-guides-code/tree/master/sign-and-combine-ecdsa/nodejs). There you can find a Node.js implementation of the code.

### Example Lit Action

> The following Lit Action uses `signAndCombineEcdsa` to combine the signatures created by each share signing the `toSign` variable, which is a transaction hash. We then serialize the transaction and sign it using `ethers.js`.

Reviewer 2: ```suggestion
The following Lit Action uses `signAndCombineEcdsa` to combine the signatures created by each share signing the `toSign` variable, which is a provided serialized transaction. We then serialize the transaction and sign it using `ethers.js`.
```

> The following Lit Action uses `signAndCombineEcdsa` to combine the signatures created by each share signing the `toSign` variable, which is the provided hash of a serialized transaction. We then serialize the transaction and sign it using `ethers.js`.

Reviewer: > The following Lit Action uses `signAndCombineEcdsa` to combine the signatures created by each share signing the `toSign` variable, which is the provided hash of a serialized transaction. We then serialize the transaction and sign it using `ethers.js`.

Imo this sentence is a little confusing. Is the transaction already serialized, and then we serialize again? Or are these two separate steps?

Also , i think we could make it more clear that each node is provisioning a share of the key which must be combined to form the complete signature

Author: Changed to:

The following Lit Action uses `signAndCombineEcdsa` to combine partial signatures provided by each node, each of which signs a share of the `toSign` variable—this variable contains the hash of the serialized transaction. After combining these shares into a complete signature, we use `ethers.js` to serialize the transaction again, this time including the signature, to finalize it for submission.

Reviewer: sounds great

> ```jsx

Reviewer 2: Idk if we should, but we could use the syntax:
```ts
const _litActionCode = async () => {
    // ...
}

export const litActionCode = `(${_litActionCode.toString())();`;
```

to provide syntax highlighting, but maybe this syntax is confusing for users. It also won't work when users try to import 3rd part libraries
```

Author: I think it's alright if we show it like this in the dev guides code, and just use strings in the docs? 

Considering this is just a starter guide, I think the current implementation is fine

Reviewer 2: It's just that the syntax highlighting is a bit broken
<img width="964" alt="image" src="https://github.com/user-attachments/assets/277def04-9613-4f1a-9a98-e08e959ed764">

(async () => {
  const signature = await Lit.Actions.signAndCombineEcdsa({
    toSign,
    publicKey,
    sigName,
  });

  const jsonSignature = JSON.parse(signature);
  jsonSignature.r = "0x" + jsonSignature.r.substring(2);
  jsonSignature.s = "0x" + jsonSignature.s;
  const hexSignature = ethers.utils.joinSignature(jsonSignature);

  const signedTx = ethers.utils.serializeTransaction(
    unsignedTransaction,
    hexSignature
  );

  const recoveredAddress = ethers.utils.recoverAddress(toSign, hexSignature);
  console.log("Recovered Address:", recoveredAddress);

  const response = await Lit.Actions.runOnce(
    { waitForResponse: true, name: "txnSender" },
    async () => {
      try {
        const rpcUrl = await Lit.Actions.getRpcUrl({ chain });
        const provider = new ethers.providers.JsonRpcProvider(rpcUrl);
        const transactionReceipt = await provider.sendTransaction(signedTx);

        return `Transaction Sent Successfully. Transaction Hash: ${transactionReceipt.hash}`;
      } catch (error) {
        return `Error: When sending transaction: ${error.message}`;
      }
    }
  );

  Lit.Actions.setResponse({ response });
})();
`
```

## Summary

> This guide demonstrates how to combine PKP signature shares, and how to used the combined shares to sign a message or transaction within a Lit Action.

Reviewer 2: ```suggestion
This guide demonstrates how to combine PKP signature shares, and submit a signed transaction all within a Lit Action.
```
